Make torch.device usable as a context manager

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91525

Fixes https://github.com/pytorch/pytorch/issues/82296
Fixes https://github.com/pytorch/pytorch/issues/27878
Fixes https://github.com/pytorch/pytorch/issues/260

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Add torch.utils.device_mode

Fixes https://github.com/pytorch/pytorch/issues/82296
Fixes https://github.com/pytorch/pytorch/issues/27878
Fixes https://github.com/pytorch/pytorch/issues/260

Open to bikeshedding the module name.  Open to bikeshedding if any of
the identifiers should get reexporting from torch top level.  Open to
bikeshedding the global setter API name / location.

Tests coming later.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add torch.utils.device_mode"


Fixes https://github.com/pytorch/pytorch/issues/82296
Fixes https://github.com/pytorch/pytorch/issues/27878
Fixes https://github.com/pytorch/pytorch/issues/260

Open to bikeshedding the module name.  Open to bikeshedding if any of
the identifiers should get reexporting from torch top level.  Open to
bikeshedding the global setter API name / location.

Tests coming later.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented on torch/signal/windows/windows.py:
```diff
@@ -378,7 +378,7 @@ def kaiser(
                        requires_grad=requires_grad)
 
-    return torch.i0(torch.sqrt(beta * beta - torch.pow(k, 2))) / torch.i0(torch.tensor(beta))
+    return torch.i0(torch.sqrt(beta * beta - torch.pow(k, 2))) / torch.i0(torch.tensor(beta, device=device))
```
shrug, kaiser test was failing until I added this line, seems harmless enough

soumith reviewed (commented):
can we make `with torch.device(...):` work?
I know `torch.device` is an object, but it'd be clean and clever to make the `torch.device` constructor also work as the context manager :D 

Also, I don't see `torch.utils.device_mode`, you seem to have changed your mind to make it `torch.utils.DeviceMode` but didn't change the title of the PR. If we go down this path, maybe I prefer something like `with torch.utils.use_device('cuda:1'):`, i.e. specifying the intent. `device_mode` to me isn't obvious what it is.

ezyang commented:
I can definitely try to make torch.device work. I will take a look.

vadimkantorov commented:
For me, DeviceContext or DefaultDeviceContext is more clear/explicit than deivce_mode. Also second on torch.device if want to be extra-brief :)

ezyang committed:
Update on "Make torch.device usable as a context manager"


Fixes https://github.com/pytorch/pytorch/issues/82296
Fixes https://github.com/pytorch/pytorch/issues/27878
Fixes https://github.com/pytorch/pytorch/issues/260

Open to bikeshedding the module name.  Open to bikeshedding if any of
the identifiers should get reexporting from torch top level.  Open to
bikeshedding the global setter API name / location.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
Now with torch.device context manager!

ezyang committed:
Update on "Make torch.device usable as a context manager"


Fixes https://github.com/pytorch/pytorch/issues/82296
Fixes https://github.com/pytorch/pytorch/issues/27878
Fixes https://github.com/pytorch/pytorch/issues/260

Open to bikeshedding the module name.  Open to bikeshedding if any of
the identifiers should get reexporting from torch top level.  Open to
bikeshedding the global setter API name / location.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make torch.device usable as a context manager"


Fixes https://github.com/pytorch/pytorch/issues/82296
Fixes https://github.com/pytorch/pytorch/issues/27878
Fixes https://github.com/pytorch/pytorch/issues/260

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

milesial commented:
Will this also work as a decorator like `no_grad`?

albanD commented on torch/__init__.py:
```diff
@@ -444,6 +445,49 @@ def is_storage(obj):
 
+_GLOBAL_DEVICE_CONTEXT = None
+
+def set_default_tensor_device(device):
```
Why not just `set_default_device()` ?

ezyang replied:
Happy to bikeshed it to that. It has tensor in the name because of `set_default_tensor_type`. But now that I look more carefully, I see it's `set_default_dtype`, so your name is better.

albanD commented on torch/utils/_device.py:
```diff
@@ -0,0 +1,62 @@
+        torch.linspace,
+        torch.logspace,
+        torch.nested.nested_tensor,
+        # torch.normal,
```
Why?

ezyang replied:
This constructor, for some godforsaken reason, doesn't actually take a device argument 🤣

albanD commented on torch/utils/_device.py:
```diff
@@ -0,0 +1,62 @@
+
+@functools.lru_cache(1)
+def _device_constructors():
+    return {
```
Feels like this can easily get out of date? :/

ezyang replied:
The OpInfo test hopefully will help; in particular, we will notice if you failed to match a constructor but it is exercised in OpInfo. But yes, the implementation strategy here requires us to know which set of Python binding functions we need to match against. For extra safety, maybe we could have codegen dump out a list of schemas that have device arguments, but you're still on the hook for making sure that other factory functions implemented purely in Python are torch function'able. Which, BTW, we also have a test for in test_overrides.py.

albanD commented on torch/utils/_device.py:
```diff
@@ -0,0 +1,62 @@
+        if func in _device_constructors():
+            if kwargs.get('device') is None:
+                kwargs['device'] = self.device
+            return func(*args, **kwargs)
```
nit: this line is not needed.

albanD commented on torch/utils/_device.py:
```diff
@@ -0,0 +1,62 @@
+        self.device = torch.device(device)
+
+    def __torch_function__(self, func, types, args=(), kwargs=None):
+        if kwargs is None:
+            kwargs = {}
```
nit
```suggestion
        kwargs = kwargs or {}
```

albanD reviewed (commented):
> Will this also work as a decorator like no_grad?

We could yes. There is no `__call__` currently on the device object.
We might want to re-use https://github.com/pytorch/pytorch/blob/master/torch/autograd/grad_mode.py#L18 as well

ezyang commented:
Ok I'll add decorator support too

ezyang commented:
I cannot easily reuse `_DecoratorContextManager` as it is written in Python but torch.device is bound entirely from C

ezyang commented:
I did a major refactor of _DecoratorContextManager to get it usable lol

ezyang committed:
Update on "Make torch.device usable as a context manager"


Fixes https://github.com/pytorch/pytorch/issues/82296
Fixes https://github.com/pytorch/pytorch/issues/27878
Fixes https://github.com/pytorch/pytorch/issues/260

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make torch.device usable as a context manager and decorator"


Fixes https://github.com/pytorch/pytorch/issues/82296
Fixes https://github.com/pytorch/pytorch/issues/27878
Fixes https://github.com/pytorch/pytorch/issues/260

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
Hmm, so I've implemented torch.device as a callable but a bit nervous about it, because I broke code that tested if an attribute of Tensor is `callable()`. Previous `x.device` is not callable, but now it is (but not in the way you expect; it's not a method!) So a little ambivalent about whether or not to support the decorator spelling.

ezyang committed:
Update on "Make torch.device usable as a context manager and decorator"


Fixes https://github.com/pytorch/pytorch/issues/82296
Fixes https://github.com/pytorch/pytorch/issues/27878
Fixes https://github.com/pytorch/pytorch/issues/260

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

albanD commented on torch/csrc/Device.cpp:
```diff
@@ -169,6 +169,36 @@ PyObject* THPDevice_reduce(PyObject* _self, PyObject* noargs) {
+
+PyObject* THPDevice_exit(PyObject* self, PyObject* unused) {
+  HANDLE_TH_ERRORS
+  at::impl::PythonTorchFunctionTLS::pop_stack();
```
Should we assert that we get a DeviceContext here?

ezyang replied:
TorchDispatchMode doesn't currently test for this:

```
    def __enter__(self):
        _push_mode(self)
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        _pop_mode()
```

I suppose we could add it but we ought to add it everywhere if we do.

albanD reviewed (approved):
I agree the device being callable could be a problem. I'm happy to drop the decorator as it is a bit dangerous.

The refactor looks good!

vadimkantorov commented:
It would also be good to have explicitly some usecases for this feature (to prevent people relying on it when they shouldn't) and forewarnings of what can go wrong. E.g. I can imagine that some idioms of using torch functions and then calling numpy on them (explicitly or implicitly) would break because e.g. `torch.arange` would suddenly placed on a strange device that `.numpy()` with `force=False` by default does not expect.

Introduction of this feature may also be a good point to reflect on the idiom of default device in general, e.g. should we encourage people to preferrably specify explicitly `device = 'cpu'` where before they would not specify the device, unless it was to be placed on GPU (or `same device as`)

ezyang commented:
I'm going to drop the callable for now. But if we drop it, there is a reasonable question about what the alternative API name should be. I'm going to do `set_device` for now but also open to bikeshedding here (and for now I'm not going to expose it from torch global namespace).

> It would also be good to have explicitly some usecases for this feature (to prevent people relying on it when they shouldn't) and forewarnings of what can go wrong. E.g. I can imagine that some idioms of using torch functions and then calling numpy on them (explicitly or implicitly) would break because e.g. torch.arange would suddenly placed on a strange device that .numpy() with force=False by default does not expect.

TBH, I'm really not sure what to say here, because I don't have enough experience using this feature to know. Anecdotally when we trialed the feature in the issue thread, people didn't noticed anything surprising (but it was also, like, three users lol). The thing where numpy (or really any custom user code) expects to get a CPU tensor but doesn't is valid but... it's exactly what the user asked for; they would have had the same problem even if they plumbed the device argument by hand. The resolution to the problem is obvious too: either you go and add an explicit `device="cpu"` argument to the factory, or you change your code to not use numpy/CPU-only code (but of course that's what you'd need to do if you wanted to run your code on GPU.) The docs already mention that explicit device arguments are respected, so I'm not sure what else to say here.

BTW, one of the reasons we didn't really want to do this context manager originally, was that the obvious implementation mechanism would have also affected C++ code as well. And then we had a testing problem, which is that all of the library code in PyTorch also would have to be indifferent to device changes, but this could actually be pretty nasty because we have a lot of CPU specific code in our CPU backend hanging about that wouldn't be invariant in this way. But we actually don't have this problem in this PR, because it's implemented using `__torch_function__`, so we only apply device arguments on the Python-to-C++ boundary (and everything in C++ has traditional semantics.) This, IMO, makes it substantially less bad.

> Introduction of this feature may also be a good point to reflect on the idiom of default device in general, e.g. should we encourage people to preferrably specify explicitly device = 'cpu' where before they would not specify the device, unless it was to be placed on GPU (or same device as)

But... not actually? As in, this feature is explicitly targeted at user code which was written in a device oblivious way, and now people want to retarget it at CUDA/MPS. If you explicitly specified devices, this feature will stop working (because we always respect explicit device arguments).

Actually, one of the things I was surprised the beta testers didn't ask for was a way to override device calls at some device; e.g., if there was a codebase that was hard-coded to only work on cuda (with explicit `device="cuda"` everywhere) and now they wanted to retarget it on a different device/MPS. This is easy enough to add, but I left it out because no one asked for it.

ezyang committed:
Update on "Make torch.device usable as a context manager and decorator"


Fixes https://github.com/pytorch/pytorch/issues/82296
Fixes https://github.com/pytorch/pytorch/issues/27878
Fixes https://github.com/pytorch/pytorch/issues/260

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

albanD reviewed (approved):
Ok!
Thanks for the update.

Skylion007 commented:
If I recall didn't Tensorflow 1.0 have this feature? Interesting that things have gone full circle in Pytorch 2.0

jbschlosser reviewed (commented):
Wow never thought I'd see the day this would be introduced! :p

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge -f "rocm timeout only"

ezyang commented:
> If I recall didn't Tensorflow 1.0 have this feature? Interesting that things have gone full circle in Pytorch 2.0

the writing was on the wall when we added torch function mode haha

mehtanirav commented:
This PR is causing ExecuTorch CI failures internally with the error:

```
ERROR: test_transformer_encode_full_dynamic (executorch.test.end2end.test_end2end.DynamicShapeMemoryPlanningE2ETest)
A TestCase method that traces/exports/tests an nn.Module and method.
...
func = cast(F, lambda *args, **kwargs: func(*args, **kwargs))
  [Previous line repeated 970 more times]
RecursionError: maximum recursion depth exceeded
``` 

sample internal [failure link](https://www.internalfb.com/intern/testinfra/diagnostics/2814749915648555.281475052894770.1672948820/)

mehtanirav commented:
@pytorchbot revert -m "Internal breakages" -c "ghfirst"

jon-chuang commented:
Oops, do we need to track this in AOTAutograd now?

Thankfully, it's unlikely to be a widely used API, given that it was only introduced recently.

albanD commented:
> Thankfully, it's unlikely to be a widely used API, given that it was only introduced recently.

It is actually used quite a bit as we recommend this is used with the "meta" or "gpu" device to speedup model creation.
But this is implemented at the torch_function mode so the behavior of this should happen when you go from python ops to to aten ops.

jon-chuang commented:
Hmm, actually, I realize that this doesn't even work like autocast, it only sets device defaults for tensor initializations.

```python
cpu_tensor = torch.ones(10, device='cpu')
with torch.device('cuda'):
    t = torch.zeros(10)
    t1 = cpu_tensor + 1

print(t.device, t1.device)  # cuda:0, cpu
```

The autocast semantic would work by coercing all new tensor _allocations_ and _incompatible operand device_ to the device, via `.to(device)`

ezyang commented:
It's a torch function mode, so it can be implemented purely in dynamo
